File: packages/nitro/ios/HybridLottieView.swift
import Lottie
import NitroModules
import UIKit

class HybridLottieView: HybridLottieViewSpec {
  private let animationView = LottieAnimationView(frame: .zero)

  var view: UIView { animationView }

  var resizeMode: ResizeMode?
  var renderMode: RenderMode?
  var sourceName: String?
  var sourceJson: String?
  var sourceURL: String?
  var sourceDotLottieURI: String?
  var imageAssetsFolder: String?
  var progress: Double?
  var speed: Double?
  var loop: Bool?
  var autoPlay: Bool?
  var enableMergePathsAndroidForKitKatAndAbove: Bool?
  var applyOpacityToLayersAndroid: Bool?
  var enableSafeModeAndroid: Bool?
  var hardwareAccelerationAndroid: Bool?
  var cacheComposition: Bool?
  var colorFilters: [LottieColorFilter]?
  var textFiltersAndroid: [TextFilterAndroid]?
  var textFiltersIOS: [TextFilterIOS]?
  var onAnimationFinish: ((_ isCancelled: Bool) -> Void)?
  var onAnimationFailure: ((_ error: String) -> Void)?
  var onAnimationLoaded: (() -> Void)?

  private struct ColorFilterSnapshot: Hashable {
    let keypath: String

    let colorBits: UInt64

    var color: Double { Double(bitPattern: colorBits) }

    init(keypath: String, color: Double) {
      self.keypath = keypath
      self.colorBits = color.bitPattern
    }
  }

  private enum SourceKind: Equatable {
    case name(String), json(String), url(String), dotLottie(String)
  }

  private var appliedDidApplyOnce = false
  private var appliedSource: SourceKind?
  private var appliedRenderMode: String?
  private var appliedResizeMode: String?
  private var appliedImageAssetsFolder = ""
  private var appliedCacheComposition = true
  private var appliedProgress: Double = 0
  private var appliedSpeed: Double = 1
  private var appliedLoop = true
  private var appliedAutoPlay = false
  private var appliedTextFilters: [String: String] = [:]
  private var appliedColorFilters: [ColorFilterSnapshot] = []
  private var installedColorKeypaths: Set<AnimationKeypath> = []

  private var loadGeneration: UInt64 = 0
  private var inFlightLoad: URLSessionDataTask?

  private var suppressFinishDepth = 0
  private var playbackEpoch: UInt64 = 0

  private var sourceJustChanged = false

  override init() {
    super.init()
    animationView.animationLoaded = { [weak self] _, _ in
      guard let self else { return }
      self.onAnimationLoaded?()
      self.applyCompositionDependent()
    }
  }

  func afterUpdate() {
    withFinishSuppressed {
      let nextSource = pendingSource()
      sourceJustChanged = nextSource != appliedSource

      appliedCacheComposition = cacheComposition ?? true

      applyImageProvider()
      applyTextProvider()

      applyRenderMode()
      applyResizeMode()

      if sourceJustChanged {
        appliedSource = nextSource
        loadGeneration &+= 1
        inFlightLoad?.cancel()
        inFlightLoad = nil
        installedColorKeypaths = []
        apply(source: nextSource)
      } else {
        applyCompositionDependent()
      }

      appliedDidApplyOnce = true
      sourceJustChanged = false
    }
  }

  func onDropView() {
    loadGeneration &+= 1
    playbackEpoch &+= 1
    suppressFinishDepth += 1
    inFlightLoad?.cancel()
    inFlightLoad = nil
    animationView.animationLoaded = nil
    for keypath in installedColorKeypaths {
      animationView.removeValueProvider(for: keypath)
    }
    installedColorKeypaths = []
    animationView.stop()
    animationView.animation = nil
    onAnimationFinish = nil
    onAnimationFailure = nil
    onAnimationLoaded = nil
  }

  private func applyImageProvider() {
    let folder = imageAssetsFolder ?? ""
    guard folder != appliedImageAssetsFolder || !appliedDidApplyOnce else { return }
    animationView.imageProvider = folder.isEmpty
      ? BundleImageProvider(bundle: .main, searchPath: nil)
      : BundleImageProvider(bundle: .main, searchPath: folder)
    appliedImageAssetsFolder = folder
  }

  private func applyTextProvider() {
    let next = pendingTextFilters()
    guard next != appliedTextFilters || !appliedDidApplyOnce else { return }
    animationView.textProvider = next.isEmpty
      ? DefaultTextProvider()
      : DictionaryTextProvider(next)
    appliedTextFilters = next
  }

  private func applyRenderMode() {
    guard let mode = renderMode else { return }
    let key = String(describing: mode)
    guard key != appliedRenderMode else { return }
    let engine: RenderingEngineOption
    switch mode {
    case .software: engine = .mainThread
    case .hardware: engine = .coreAnimation
    case .automatic: engine = .automatic
    }
    animationView.configuration = LottieConfiguration(renderingEngine: engine)
    appliedRenderMode = key
  }

  private func applyResizeMode() {
    guard let mode = resizeMode else { return }
    let key = String(describing: mode)
    guard key != appliedResizeMode else { return }
    switch mode {
    case .cover: animationView.contentMode = .scaleAspectFill
    case .contain: animationView.contentMode = .scaleAspectFit
    case .center: animationView.contentMode = .center
    }
    appliedResizeMode = key
  }

  private func pendingSource() -> SourceKind? {
    if let s = sourceJson, !s.isEmpty { return .json(s) }
    if let s = sourceName, !s.isEmpty { return .name(s) }
    if let s = sourceDotLottieURI, !s.isEmpty { return .dotLottie(s) }
    if let s = sourceURL, !s.isEmpty { return .url(s) }
    return nil
  }

  private func resolveURL(_ raw: String) -> URL? {
    if let url = URL(string: raw), url.scheme != nil { return url }
    return URL(fileURLWithPath: raw, relativeTo: Bundle.main.resourceURL)
  }

  private func apply(source: SourceKind?) {
    let generation = loadGeneration
    let cache = appliedCacheComposition ? LottieAnimationCache.shared : nil
    let dotCache = appliedCacheComposition ? DotLottieCache.sharedCache : nil

    switch source {
    case .none:
      animationView.animation = nil

    case .json(let json):
      guard let data = json.data(using: .utf8),
            let animation = try? LottieAnimation.from(data: data) else {
        emitFailure("Unable to create the lottie animation object from the JSON source")
        return
      }
      animationView.animation = animation

    case .name(let name):
      guard let animation = LottieAnimation.named(name, bundle: .main, animationCache: cache) else {
        emitFailure("Unable to find the lottie animation named \"\(name)\" in the main bundle")
        return
      }
      animationView.animation = animation

    case .url(let raw):
      guard let url = resolveURL(raw) else {
        emitFailure("Invalid Lottie animation URL: \(raw)")
        return
      }
      animationView.animation = nil
      inFlightLoad = LottieAnimation.loadedFrom(url: url, closure: { [weak self] animation in
        guard let self, generation == self.loadGeneration else { return }
        self.inFlightLoad = nil
        guard let animation else {
          self.emitFailure("Unable to fetch the Lottie animation from the URL: \(url.absoluteString)")
          return
        }
        self.animationView.animation = animation
      }, animationCache: cache)

    case .dotLottie(let raw):
      guard let url = resolveURL(raw) else {
        emitFailure("Invalid .lottie URL: \(raw)")
        return
      }
      animationView.animation = nil
      inFlightLoad = DotLottieFile.loadedFrom(url: url, dotLottieCache: dotCache) { [weak self] result in
        guard let self, generation == self.loadGeneration else { return }
        self.inFlightLoad = nil
        switch result {
        case .failure(let error):
          self.emitFailure(error.localizedDescription)
        case .success(let file):
          self.animationView.loadAnimation(nil, from: file)
        }
      }
    }
  }

  private func applyCompositionDependent() {
    guard animationView.animation != nil else { return }
    withFinishSuppressed {
      appliedLoop = loop ?? true
      appliedSpeed = speed ?? 1
      animationView.loopMode = appliedLoop ? .loop : .playOnce
      animationView.animationSpeed = CGFloat(appliedSpeed)

      reconcileColorFilters()
      applyPlayback()
    }
  }

  private func applyPlayback() {
    let wantsPlay = autoPlay ?? false
    let next = progress ?? 0
    let progressChanged = next != appliedProgress
    appliedProgress = next

    if wantsPlay {
      let shouldStart = !appliedAutoPlay
        || progressChanged
        || sourceJustChanged
        || !animationView.isAnimationPlaying
      if shouldStart {
        let from = next >= 1 ? 0 : next
        animationView.play(
          fromProgress: AnimationProgressTime(from),
          toProgress: 1,
          loopMode: appliedLoop ? .loop : .playOnce,
          completion: makeFinishCompletion()
        )
      }
    } else if progressChanged {
      animationView.currentProgress = AnimationProgressTime(next)
    }
    appliedAutoPlay = wantsPlay
  }

  private func reconcileColorFilters() {
    let next = (colorFilters ?? []).map {
      ColorFilterSnapshot(keypath: $0.keypath, color: $0.color)
    }
    guard next != appliedColorFilters || sourceJustChanged || !appliedDidApplyOnce else { return }

    var desired: [AnimationKeypath: UIColor] = [:]
    for filter in next {
      guard let color = Self.colorFromARGB(filter.color) else {
        emitFailure(
          "Unresolvable colorFilters color for keypath \"\(filter.keypath)\""
        )
        continue
      }
      desired[AnimationKeypath(keypath: "\(filter.keypath).**.Color")] = color
    }

    for stale in installedColorKeypaths.subtracting(desired.keys) {
      animationView.removeValueProvider(for: stale)
    }
    for (keypath, color) in desired {
      animationView.setValueProvider(ColorValueProvider(color.lottieColorValue), keypath: keypath)
    }
    installedColorKeypaths = Set(desired.keys)
    appliedColorFilters = next
  }

  private static func colorFromARGB(_ value: Double) -> UIColor? {
    guard value >= -2_147_483_648, value <= 4_294_967_295 else { return nil }
    let argb = UInt32(truncatingIfNeeded: Int64(value))
    return UIColor(
      red: CGFloat((argb >> 16) & 0xFF) / 255,
      green: CGFloat((argb >> 8) & 0xFF) / 255,
      blue: CGFloat(argb & 0xFF) / 255,
      alpha: CGFloat((argb >> 24) & 0xFF) / 255
    )
  }

  private func makeFinishCompletion() -> LottieCompletionBlock {
    playbackEpoch &+= 1
    let epoch = playbackEpoch
    return { [weak self] finished in
      guard let self,
            epoch == self.playbackEpoch,
            self.suppressFinishDepth == 0 else { return }
      self.playbackEpoch &+= 1
      self.onAnimationFinish?(!finished)
    }
  }

  private func withFinishSuppressed(_ body: () -> Void) {
    suppressFinishDepth += 1
    defer { suppressFinishDepth -= 1 }
    body()
  }

  private func emitFailure(_ message: String) {
    onAnimationFailure?(message)
  }

  private func pendingTextFilters() -> [String: String] {
    Dictionary(
      (textFiltersIOS ?? []).map { ($0.keypath, $0.text) },
      uniquingKeysWith: { _, last in last }
    )
  }

  func play(startFrame: Double, endFrame: Double) throws {
  }

  func reset() throws {}
  func pause() throws {}
  func resume() throws {}
}

File: packages/nitro/android/src/main/java/com/margelo/nitro/lottienitro/HybridLottieView.kt
package com.margelo.nitro.lottienitro

import android.animation.Animator
import android.graphics.ColorFilter
import android.graphics.Typeface
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.view.View
import android.widget.ImageView
import androidx.annotation.Keep
import com.airbnb.lottie.FontAssetDelegate
import com.airbnb.lottie.LottieAnimationView
import com.airbnb.lottie.LottieDrawable
import com.airbnb.lottie.LottieProperty
import com.airbnb.lottie.RenderMode as LottieRenderMode
import com.airbnb.lottie.SimpleColorFilter
import com.airbnb.lottie.TextDelegate
import com.airbnb.lottie.model.KeyPath
import com.airbnb.lottie.value.LottieValueCallback
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.views.text.ReactFontManager
import java.io.File
import java.io.FileInputStream
import java.util.regex.Pattern
import java.util.zip.ZipInputStream

@Keep
@DoNotStrip
class HybridLottieView(
  val context: ThemedReactContext,
) : HybridLottieViewSpec() {
  private val animationView = LottieAnimationView(context).apply {
    scaleType = ImageView.ScaleType.CENTER_INSIDE
  }

  override val view: View = animationView

  override var resizeMode: ResizeMode? = null
  override var renderMode: RenderMode? = null
  override var sourceName: String? = null
  override var sourceJson: String? = null
  override var sourceURL: String? = null
  override var sourceDotLottieURI: String? = null
  override var imageAssetsFolder: String? = null
  override var progress: Double? = null
  override var speed: Double? = null
  override var loop: Boolean? = null
  override var autoPlay: Boolean? = null
  override var enableMergePathsAndroidForKitKatAndAbove: Boolean? = null
  override var applyOpacityToLayersAndroid: Boolean? = null
  override var enableSafeModeAndroid: Boolean? = null
  override var hardwareAccelerationAndroid: Boolean? = null
  override var cacheComposition: Boolean? = null
  override var colorFilters: Array<LottieColorFilter>? = null
  override var textFiltersAndroid: Array<TextFilterAndroid>? = null
  override var textFiltersIOS: Array<TextFilterIOS>? = null
  override var onAnimationFinish: ((isCancelled: Boolean) -> Unit)? = null
  override var onAnimationFailure: ((error: String) -> Unit)? = null
  override var onAnimationLoaded: (() -> Unit)? = null

  private sealed interface SourceKind {
    data class Name(val v: String) : SourceKind
    data class Json(val v: String) : SourceKind
    data class Url(val v: String) : SourceKind
    data class DotLottie(val v: String) : SourceKind
  }

  private data class ColorFilterSnapshot(val keypath: String, val colorBits: Long)

  private var appliedDidApplyOnce = false
  private var appliedSource: SourceKind? = null
  private var appliedRenderMode: LottieRenderMode? = null
  private var appliedScaleType: ImageView.ScaleType? = null
  private var appliedImageAssetsFolder = ""
  private var appliedCacheComposition = true
  private var appliedProgress = 0f
  private var appliedSpeed = 1f
  private var appliedLoop = true
  private var appliedAutoPlay = false
  private var appliedLayerType: Int? = null
  private var appliedEnableMergePaths: Boolean? = null
  private var appliedApplyOpacityToLayers: Boolean? = null
  private var appliedEnableSafeMode: Boolean? = null
  private var appliedTextFilters: List<TextFilterAndroid> = emptyList()
  private var appliedColorFilters: List<ColorFilterSnapshot> = emptyList()
  private var installedColorKeyPaths: List<KeyPath> = emptyList()

  private var loadGeneration = 0L

  private var suppressFinishDepth = 0
  private var finishEmittedForRun = false
  private var sourceJustChanged = false

  private var pendingAttachPlay: View.OnAttachStateChangeListener? = null

  private val animatorListener = object : Animator.AnimatorListener {
    override fun onAnimationStart(animation: Animator) {
      finishEmittedForRun = false
    }

    override fun onAnimationCancel(animation: Animator) = emitFinish(true)
    override fun onAnimationEnd(animation: Animator) = emitFinish(false)
    override fun onAnimationRepeat(animation: Animator) = Unit
  }

  init {
    animationView.addAnimatorListener(animatorListener)
    animationView.setFailureListener { emitFailure(it.message ?: it.toString()) }
    animationView.addLottieOnCompositionLoadedListener {
      onAnimationLoaded?.invoke()
      applyCompositionDependent()
    }
    animationView.setFontAssetDelegate(object : FontAssetDelegate() {
      override fun fetchFont(fontFamily: String): Typeface =
        ReactFontManager.getInstance().getTypeface(fontFamily, UNSET, UNSET, context.assets)

      override fun fetchFont(fontFamily: String, fontStyle: String, fontName: String): Typeface {
        val weight = when (fontStyle) {
          "Thin" -> 100
          "Light" -> 200
          "Normal", "Regular" -> 400
          "Medium" -> 500
          "Bold" -> 700
          "Black" -> 900
          else -> UNSET
        }
        return ReactFontManager.getInstance()
          .getTypeface(fontName, UNSET, weight, context.assets)
      }
    })
  }

  override fun afterUpdate() {
    withFinishSuppressed {
      val nextSource = pendingSource()
      sourceJustChanged = nextSource != appliedSource

      (cacheComposition ?: true).let {
        if (it != appliedCacheComposition || !appliedDidApplyOnce) {
          animationView.setCacheComposition(it)
          appliedCacheComposition = it
        }
      }

      (imageAssetsFolder ?: "").let {
        if (it != appliedImageAssetsFolder) {
          animationView.imageAssetsFolder = it
          appliedImageAssetsFolder = it
        }
      }

      applyIndependentFlags()
      applyTextFilters()

      if (sourceJustChanged) {
        appliedSource = nextSource
        loadGeneration++
        installedColorKeyPaths = emptyList()
        applySource(nextSource)
      } else if (animationView.composition != null) {
        applyCompositionDependent()
      }

      appliedDidApplyOnce = true
      sourceJustChanged = false
    }
  }

  override fun onDropView() {
    loadGeneration++
    suppressFinishDepth++
    animationView.removeAnimatorListener(animatorListener)
    animationView.removeAllLottieOnCompositionLoadedListener()
    animationView.setFailureListener(null)
    installedColorKeyPaths.forEach {
      animationView.clearValueCallback(it, LottieProperty.COLOR_FILTER)
    }
    installedColorKeyPaths = emptyList()
    pendingAttachPlay?.let { animationView.removeOnAttachStateChangeListener(it) }
    pendingAttachPlay = null
    animationView.setTextDelegate(null)
    animationView.cancelAnimation()
    animationView.clearAnimation()
    onAnimationFinish = null
    onAnimationFailure = null
    onAnimationLoaded = null
  }

  private fun applyIndependentFlags() {
    resizeMode?.let { mode ->
      val scaleType = when (mode) {
        ResizeMode.COVER -> ImageView.ScaleType.CENTER_CROP
        ResizeMode.CONTAIN -> ImageView.ScaleType.FIT_CENTER
        ResizeMode.CENTER -> ImageView.ScaleType.CENTER_INSIDE
      }
      if (scaleType != appliedScaleType) {
        animationView.scaleType = scaleType
        appliedScaleType = scaleType
      }
    }

    renderMode?.let { mode ->
      val native = when (mode) {
        RenderMode.AUTOMATIC -> LottieRenderMode.AUTOMATIC
        RenderMode.HARDWARE -> LottieRenderMode.HARDWARE
        RenderMode.SOFTWARE -> LottieRenderMode.SOFTWARE
      }
      if (native != appliedRenderMode) {
        animationView.renderMode = native
        appliedRenderMode = native
      }
    }

    hardwareAccelerationAndroid?.let {
      val layerType = if (it) View.LAYER_TYPE_HARDWARE else View.LAYER_TYPE_SOFTWARE
      if (layerType != appliedLayerType) {
        animationView.setLayerType(layerType, null)
        appliedLayerType = layerType
      }
    }

    enableMergePathsAndroidForKitKatAndAbove?.let {
      if (it != appliedEnableMergePaths) {
        animationView.enableMergePathsForKitKatAndAbove(it)
        appliedEnableMergePaths = it
      }
    }
    applyOpacityToLayersAndroid?.let {
      if (it != appliedApplyOpacityToLayers) {
        animationView.setApplyingOpacityToLayersEnabled(it)
        appliedApplyOpacityToLayers = it
      }
    }
    enableSafeModeAndroid?.let {
      if (it != appliedEnableSafeMode) {
        animationView.setSafeMode(it)
        appliedEnableSafeMode = it
      }
    }
  }

  private fun applyTextFilters() {
    val next = textFiltersAndroid?.toList() ?: emptyList()
    if (next == appliedTextFilters && appliedDidApplyOnce) return

    if (next.isEmpty()) {
      if (appliedTextFilters.isNotEmpty()) {
        animationView.setTextDelegate(null)
        animationView.invalidate()
      }
    } else {
      animationView.setTextDelegate(
        TextDelegate(animationView).apply {
          next.forEach { setText(it.find, it.replace) }
        }
      )
    }
    appliedTextFilters = next
  }

  private fun pendingSource(): SourceKind? {
    sourceJson?.takeIf { it.isNotEmpty() }?.let { return SourceKind.Json(it) }
    sourceName?.takeIf { it.isNotEmpty() }?.let { return SourceKind.Name(it) }
    sourceDotLottieURI?.takeIf { it.isNotEmpty() }?.let { return SourceKind.DotLottie(it) }
    sourceURL?.takeIf { it.isNotEmpty() }?.let { return SourceKind.Url(it) }
    return null
  }

  private fun applySource(source: SourceKind?) {
    val generation = loadGeneration
    fun key(s: String): String? = if (appliedCacheComposition) s.hashCode().toString() else null

    try {
      when (source) {
        null -> animationView.setAnimation(null as String?)

        is SourceKind.Json -> animationView.setAnimationFromJson(source.v, key(source.v))

        is SourceKind.Name -> {
          val name = if (source.v.contains(".")) source.v else "${source.v}.json"
          animationView.setAnimation(name)
        }

        is SourceKind.Url -> {
          val file = File(source.v)
          if (file.exists()) {
            animationView.setAnimation(FileInputStream(file), key(source.v))
          } else {
            animationView.setAnimationFromUrl(source.v)
          }
        }

        is SourceKind.DotLottie -> applyDotLottie(source.v)
      }
    } catch (t: Throwable) {
      if (generation == loadGeneration) emitFailure(t.message ?: t.toString())
    }
  }

  private fun applyDotLottie(assetName: String) {
    val file = File(assetName)
    if (file.exists()) {
      animationView.setAnimation(
        ZipInputStream(FileInputStream(file)),
        if (appliedCacheComposition) assetName.hashCode().toString() else null,
      )
      return
    }

    val scheme = runCatching { Uri.parse(assetName).scheme }.getOrNull()
    if (scheme != null) {
      if (scheme == "file") {
        val path = Uri.parse(assetName).path
        if (path == null) {
          emitFailure("URI path is null for .lottie asset: $assetName")
          return
        }
        val fileWithScheme = File(path)
        if (!fileWithScheme.exists()) {
          emitFailure("The .lottie file does not exist: $path")
          return
        }
        animationView.setAnimation(
          ZipInputStream(FileInputStream(fileWithScheme)),
          if (appliedCacheComposition) assetName.hashCode().toString() else null,
        )
      } else {
        animationView.setAnimationFromUrl(assetName)
      }
      return
    }

    val resourceId = animationView.resources.getIdentifier(
      assetName, "raw", context.packageName,
    )
    if (resourceId == 0) {
      emitFailure("Animation for $assetName was not found in raw resources")
      return
    }
    animationView.setAnimation(resourceId)
  }

  private fun applyCompositionDependent() {
    if (animationView.composition == null) return
    withFinishSuppressed {
      (loop ?: true).let {
        if (it != appliedLoop || sourceJustChanged || !appliedDidApplyOnce) {
          animationView.repeatCount = if (it) LottieDrawable.INFINITE else 0
          appliedLoop = it
        }
      }
      (speed ?: 1.0).toFloat().let {
        if (it != appliedSpeed || sourceJustChanged || !appliedDidApplyOnce) {
          animationView.speed = it
          appliedSpeed = it
        }
      }

      reconcileColorFilters()
      applyPlayback()
    }
  }

  private fun applyPlayback() {
    val wantsPlay = autoPlay ?: false
    val next = (progress ?: 0.0).toFloat()
    val progressChanged = next != appliedProgress

    if (progressChanged) {
      animationView.progress = next
      appliedProgress = next
    }

    if (wantsPlay) {
      val shouldStart = !appliedAutoPlay ||
        progressChanged ||
        sourceJustChanged ||
        !animationView.isAnimating
      if (shouldStart) {
        finishEmittedForRun = false
        animationView.resumeAnimation()
      }
    }
    appliedAutoPlay = wantsPlay
  }

  private fun reconcileColorFilters() {
    val next = (colorFilters ?: emptyArray()).map {
      ColorFilterSnapshot(it.keypath, it.color.toRawBits())
    }
    if (next == appliedColorFilters && !sourceJustChanged && appliedDidApplyOnce) return

    val desired = mutableListOf<Pair<KeyPath, Int>>()
    for (filter in next) {
      val color = colorFromARGB(Double.fromBits(filter.colorBits))
      if (color == null) {
        emitFailure(
          "Unresolvable colorFilters color for keypath \"${filter.keypath}\""
        )
        continue
      }
      val keys = "${filter.keypath}.**"
        .split(Pattern.quote(".").toRegex())
        .dropLastWhile { it.isEmpty() }
        .toTypedArray()
      desired.add(KeyPath(*keys) to color)
    }

    val desiredKeyPaths = desired.map { it.first }
    (installedColorKeyPaths - desiredKeyPaths.toSet()).forEach {
      animationView.clearValueCallback(it, LottieProperty.COLOR_FILTER)
    }
    desired.forEach { (keyPath, color) ->
      animationView.addValueCallback(
        keyPath,
        LottieProperty.COLOR_FILTER,
        LottieValueCallback<ColorFilter>(SimpleColorFilter(color)),
      )
    }
    installedColorKeyPaths = desiredKeyPaths
    appliedColorFilters = next
  }

  private fun colorFromARGB(value: Double): Int? {
    if (value < -2_147_483_648.0 || value > 4_294_967_295.0) return null
    return value.toLong().toInt()
  }

  private fun emitFinish(cancelled: Boolean) {
    if (suppressFinishDepth > 0) return
    if (finishEmittedForRun) return
    finishEmittedForRun = true
    onAnimationFinish?.invoke(cancelled)
  }

  private fun emitFailure(message: String) {
    onAnimationFailure?.invoke(message)
  }

  private inline fun withFinishSuppressed(body: () -> Unit) {
    suppressFinishDepth++
    try {
      body()
    } finally {
      suppressFinishDepth--
    }
  }

  override fun play(startFrame: Double, endFrame: Double) {
    val hasRange = startFrame != -1.0 && endFrame != -1.0
    onUiThread {
      if (hasRange) {
        animationView.setMinAndMaxFrame(startFrame.toInt(), endFrame.toInt())
      } else {
        val composition = animationView.composition
        if (composition != null) {
          val start = composition.startFrame.toInt()
          val end = composition.endFrame.toInt()
          if (animationView.minFrame.toInt() != start || animationView.maxFrame.toInt() != end) {
            animationView.setMinAndMaxFrame(start, end)
          }
        }
      }

      whenAttached {
        finishEmittedForRun = false
        if (hasRange) {
          animationView.playAnimation()
        } else {
          animationView.resumeAnimation()
        }
      }
    }
  }

  override fun reset() {
    onUiThread {
      animationView.cancelAnimation()
      animationView.progress = 0f
      appliedProgress = 0f
    }
  }

  override fun pause() {
    onUiThread {
      animationView.pauseAnimation()
      emitFinish(true)
    }
  }

  override fun resume() {
    onUiThread {
      finishEmittedForRun = false
      animationView.resumeAnimation()
    }
  }

  private inline fun onUiThread(crossinline body: () -> Unit) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
      body()
    } else {
      Handler(Looper.getMainLooper()).post { body() }
    }
  }

  private fun whenAttached(body: () -> Unit) {
    if (animationView.isAttachedToWindow) {
      body()
      return
    }
    pendingAttachPlay?.let { animationView.removeOnAttachStateChangeListener(it) }
    val listener = object : View.OnAttachStateChangeListener {
      override fun onViewAttachedToWindow(v: View) {
        animationView.removeOnAttachStateChangeListener(this)
        pendingAttachPlay = null
        body()
      }

      override fun onViewDetachedFromWindow(v: View) {
        animationView.removeOnAttachStateChangeListener(this)
        pendingAttachPlay = null
      }
    }
    pendingAttachPlay = listener
    animationView.addOnAttachStateChangeListener(listener)
  }

  private companion object {
    const val UNSET = -1
  }
}
